library('shiny')

shinyUI(navbarPage(
    title = 'ReCount2: analysis-ready RNA-seq gene and exon counts datasets',
    tabPanel('Datasets',
        h1('ReCount2'),
        h3('A multi-experiment resource of analysis-ready RNA-seq gene and exon count datasets'),
        p(strong('ReCount2'), ' is an online resource consisting of RNA-seq gene and exon counts as well as coverage bigWig files for 2039 different studies. It is the second generation of the ', HTML('<a href=http://bowtie-bio.sourceforge.net/recount/>ReCount project</a>.'), ' The raw sequencing data were processed with ', HTML('<a href="http://rail.bio">Rail-RNA</a>'), ' as described at ', HTML('<a href="http://biorxiv.org/content/early/2016/01/29/038224">bioRxiv 038224</a>'), ' which created the coverage bigWig files. For ease of statistical analysis, for each study we created count tables at the gene and exon levels and extracted phenotype data, which we provide in their raw formats as well as in RangedSummarizedExperiment R objects (described in the ', HTML('<a href="http://bioconductor.org/packages/SummarizedExperiment">SummarizedExperiment</a>'), ' Bioconductor package). We also computed the mean coverage per study and provide it in a bigWig file, which can be used with the ', HTML('<a href="http://bioconductor.org/packages/derfinder">derfinder</a>'), ' Bioconductor package to perform annotation-agnostic differential expression analysis at the expressed regions-level as described at ', HTML('<a href="http://biorxiv.org/content/early/2015/02/19/015370">bioRxiv 015370</a>.'), ' The count tables, RangedSummarizeExperiment objects, phenotype tables, sample bigWigs, mean bigWigs, and file information tables are ready to use and freely available here. We also created the ',  HTML('<a href="http://bioconductor.org/packages/recount">recount</a>'), ' Bioconductor package which allows you to search and download the data for a specific study. By taking care of several preprocessing steps and combining many datasets into one easily-accessible website, we make finding and analyzing RNA-seq data considerably more straightforward.'),
        tags$hr(),
        h3('Related publications'),
        HTML('<b>Collado-Torres L</b>, <b>Nellore A</b>, Langmead B, Leek JT. <a href="">ReCount2</a>. <i>Journal COMPLETE ME</i> <b>Volume</b> pages.'),
        tags$hr(),
        h3('The Datasets'),
        DT::dataTableOutput('metadata'),
        tags$hr(),
        h3('Authors'),
        tags$ul(
            tags$li(HTML('<a href="http://lcolladotor.github.io/about.html">Leonardo Collado-Torres</a>')),
            tags$li(HTML('<a href="http://nellore.bio/">Abhinav Nellore</a>')),
            tags$li(HTML('<a href="http://www.cs.jhu.edu/~langmea/">Ben Langmead</a>')),
            tags$li(HTML('<a href="http://jtleek.com/">Jeff Leek</a>'))
        ),
        tags$hr(),
        p('This research was supported in part by COMPLETE ME. LCT was supported by Consejo Nacional de Ciencia y Tecnología México 351535.')
    ),
    tabPanel('Popular datasets',
        DT::dataTableOutput('popular'),
        tags$hr(),
        p('This research was supported in part by COMPLETE ME. LCT was supported by Consejo Nacional de Ciencia y Tecnología México 351535.')
    ),
    tabPanel('Documentation',
        p('All columns of the table below are sortable and searchable. The columns are as follows:'),
        h5('accession'),
        p('The SRA accession identifier for the study. The link points to SRA for a full description of the study.'),
        h5('number of samples'),
        p('The total number of samples available for the given study. Note that in some exceptional cases not all samples for a given study were analyzed with ', HTML('<a href="http://rail.bio">Rail-RNA</a>.')),
        h5('species'),
        p('The species of the samples under study.'),
        h5('abstract'),
        p('The abstract describing the study as provided by SRA via the ', HTML('<a href="http://bioconductor.org/packages/SRAdb">SRAdb</a>'), ' Bioconductor package.'),
        h5('RSE gene'),
        p('The RangedSummarizedExperiment object for the counts summarized at the gene level using the UCSC hg38 knownGene annotation as provided ', HTML('<a href="https://lcolladotor.shinyapps.io/recount2/ucsc-knowngene-hg38-genes-bp-length.Rdata">here</a>. Note that the GRanges object includes bp_length as a metadata column which is the sum of the exon widths. If two exons are overlapping, the overlapped bases are only counted once.')),
        h5('RSE exon'),
        p('The RangedSummarizedExperiment object for the counts summarized at the exon level using the UCSC hg38 knownGene annotation as provided ', HTML('<a href="https://lcolladotor.shinyapps.io/recount2/ucsc-knowngene-hg38-exons.Rdata">here</a>. This GRangesList object has 1 element per gene. For each gene this object contains the reduced exons such that they are non-overlapping within a gene.')),
        h5('counts gene'),
        p('A tsv file with the count matrix used to create the RangedSummarizedExperiment object at the gene level.'),
        h5('counts exon'),
        p('A tsv file with the count matrix used to create the RangedSummarizedExperiment object at the exon level.'),
        h5('phenotype'),
        p('The phenotype information (sample metadata) in a tsv file used for both RangedSummarizedExperiment objects. The table includes the SRA study id, the SRA sample id, the SRA experiment id, the SRA run id, the reads counts as reported by SRA, the number of reads aligned, the proportion of reads reported by SRA that aligned, whether the sample was paired-end or not, whether we think SRA misreported the paired-end label, the mapped read count, the coverage AUC, the SHARQ tissue, the SHARQ cell type, the biosample sumbmission date, the biosample publication date, the biosample update date, the average read length, and the name of the coverage bigWig file.'),
        h5('files info'),
        p('A tsv file with the names of the uploaded files, the md5sum, the size in bytes and the url to download the actual files.'),
        tags$hr(),
        h5('Getting started with RangedSummarizedExperiment objects'),
        p('Check out the ', HTML('<a href="http://bioconductor.org/packages/SummarizedExperiment">SummarizedExperiment</a>'), ' Bioconductor package for an overview on RangedSummarizedExperiment objects. Most RNA-seq differential expression Bioconductor packages use them. In particular, the vignette of the ', HTML('<a href="http://bioconductor.org/packages/DESeq2">DESeq2</a>'), ' Bioconductor package shows in detail how to perform an analysis with them. You might also want to use the ', HTML('<a href="http://bioconductor.org/packages/DEFormats">DEFormats</a>'), ' Bioconductor package for converting these objects to other formats.'),
        h5('Code used'),
        HTML("The code for analyzing the <a href='http://rail.bio'>Rail-RNA</a> output and creating this website is available via <a href='https://github.com/nellore/runs'>GitHub</a>."),
        tags$hr(),
        p('This research was supported in part by COMPLETE ME. LCT was supported by Consejo Nacional de Ciencia y Tecnología México 351535.')
    ),
    tabPanel('How to download with recount',
        p('The following R code shows how to use the', HTML('<a href="http://bioconductor.org/packages/recount">recount</a>'), 'Bioconductor package for downloading data. In this example we will download the data for study DRP000499.'),
        includeMarkdown('download.md'),
        tags$hr(),
        p('This research was supported in part by COMPLETE ME. LCT was supported by Consejo Nacional de Ciencia y Tecnología México 351535.')
    )
))